### ISSUE TITLE
Lexical mode problem with sentinel character?

### ISSUE THREAD

**Issue by RedTailedHawk** [antlr/antlr4#982]
The Islands in the Stream section in the ANTLR book discusses lexical mode switching and uses XML as an example.  The lexer has the following:

````
lexer grammar ModeTagsLexer;

// Default mode rules (the SEA)
OPEN  : '<'     -> mode(ISLAND) ;       // switch to ISLAND mode
TEXT  : ~'<'+ ;                         // clump all text together

mode ISLAND;
CLOSE : '>'     -> mode(DEFAULT_MODE) ; // back to SEA mode 
SLASH : '/' ;
ID    : [a-zA-Z]+ ;                     // match/send ID in tag to parser
````

The TEXT lexer rule includes all text *except* the < character.  But what if your text inside the XML needs to have that character?  Is this possible?

**Comment by ericvergnaud** [comment]
Within a tag this would be illegal, outside a tag it would have be represented by &lt;

Envoyé de mon iPhone

**Comment by RedTailedHawk** [comment]
For the language I'm developing, I need to support embedded SQL (which is not part of the grammar).

My lexer grammar has this:

````
RQL_KW_SQL_START : 'FROM SQL START'  -> pushMode(SQL) ;

mode SQL ; 
RQL_KW_SQL_END : '~SQL END'  -> popMode;
EMBEDDED_SQL : ~[~]+ ;
````

This allows me to do the following:

````
@out sales LABEL "The sales"
FROM SQL START
    SELECT sales
    FROM my_table
~SQL END
;
````

But the problem is that a tilde (~) cannot appear anywhere in the embedded SQL that the user enters since the EMBEDDED_SQL lexer rule has ~[~]+;

So the following generates a parsing error:

````
@out sales LABEL "~The sales~"   // Tilde is allowed here
FROM SQL START
    -- Some SQL comment with a ~ in it    // But tilde here generates parser error
    SELECT sales
    FROM my_table
~SQL END
;
````

Is it possible to define EMBEDDED_SQL so it accepts any sequence of characters except  "~SQL END"?

Thanks.

**Comment by jimidle** [comment]
Verbose token sequences and tokens including spaces are not a good idea.
You probably don't even need modes for this, how about:

SQL: 'SQL' .*? 'SQLEND' ;

Then you can process that text as you like. While discussing syntax is
perhaps the very definition of bikeshedding, it looks like you are perhaps
just designing it as you go. You will generally end up with something as
bad as SQL and as verbose as COBOL on that path.

Though you can use SQL START etc, it looks very cumbersome to me. Also, how
do you then cater for

FROM
   SQL START

and

FROM
SQL
// SQL blah blah
START

and so on.

Jim

**Comment by RedTailedHawk** [comment]
Hi Jim, thanks for the feedback.  So if I use the lexer rule as you suggest, how would my visitor extract the actual SQL?

SQL: 'SQL' .*? 'SQLEND' ;

Wouldn't I need to ignore the first 3 characters (SQL), and the last 6 (SQLEND)?  That's not a big deal to handle, but just wanted to ensure I'm understanding.  Seems a lot easier than having to deal with lexical modes.

-Larry

**Comment by jimidle** [comment]
Yes - you need to cater for the SQL and SQLEND yourself, but that is
trivial.

**Comment by RedTailedHawk** [comment]
Thanks Jim, I switched over to using the SQL token discussed above.  Works great, and simplifies things since I don't have to use lexical modes.

As for my original question on this thread, @ericvergnaud clarified above that there are limitations as to what can be used within the context of each mode.

So I think we can close this "issue".  Thanks.  :-)

**Comment by SecZhujun** [comment]
Hi! I've used the method @jimidle provided,but the '.' wildcard didn't match simbols like '(' 、 ',' 、'*'...an so on.
here's my grammar:
grammar IfElse;

prop:stat+;
stat:'IF' expression 'THEN' sql;
sql:SQL    ';';
expression:ID OP DIGIT;

SQL:'UPDATE';
ID:([a-zA-Z]|'_'|'.')+;
OP:'>';
DIGIT:[0-9];
WS:[ \r\n\t]  ->skip;
